nexus/saga-recovery/src/status.rs

```diff
@@ -0,0 +1,175 @@
+pub struct Report {
+    pub recent_recoveries: DebuggingHistory<RecoverySuccess>,
+    pub recent_failures: DebuggingHistory<RecoveryFailure>,
+    pub last_pass: LastPass,
+
+    pub ntotal_recovered: usize,
+    pub ntotal_failures: usize,
+    pub ntotal_started: usize,
+    pub ntotal_finished: usize,
+    pub ntotal_sec_errors_missing: usize,
+    pub ntotal_sec_errors_bad_state: usize,
+}
+
+impl Report {
+    pub fn new() -> Report {
+        Report {
+            recent_recoveries: DebuggingHistory::new(N_SUCCESS_SAGA_HISTORY),
+            recent_failures: DebuggingHistory::new(N_FAILED_SAGA_HISTORY),
+            last_pass: LastPass::NeverStarted,
+            ntotal_recovered: 0,
+            ntotal_failures: 0,
+            ntotal_started: 0,
+            ntotal_finished: 0,
+            ntotal_sec_errors_missing: 0,
+            ntotal_sec_errors_bad_state: 0,
+        }
+    }
+
+    /// Update the report after a single saga recovery pass where we at least
+    /// successfully constructed a plan
+    pub fn update_after_pass(
+        &mut self,
+        plan: &recovery::Plan,
+        execution: recovery::Execution,
+        nstarted: usize,
+    ) {
+        self.last_pass =
+            LastPass::Success(LastPassSuccess::new(plan, &execution));
+
+        let (succeeded, failed) = execution.into_results();
+
+        for success in succeeded {
+            self.recent_recoveries.append(success);
+            self.ntotal_recovered += 1;
+        }
+
+        for failure in failed {
+            self.recent_failures.append(failure);
+            self.ntotal_failures += 1;
+        }
+
+        self.ntotal_started += nstarted;
+        self.ntotal_finished += plan.ninferred_done();
+    }
+
+    /// Update the report after a saga recovery pass where we couldn't even
+    /// construct a plan (usually because we couldn't load state from the
+    /// database)
+    pub fn update_after_failure(&mut self, error: &Error, nstarted: usize) {
+        self.ntotal_started += nstarted;
+        self.last_pass = LastPass::Failed {
+            message: InlineErrorChain::new(error).to_string(),
+        };
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+pub struct RecoverySuccess {
+    pub time: DateTime<Utc>,
+    pub saga_id: SagaId,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+pub struct RecoveryFailure {
+    pub time: DateTime<Utc>,
+    pub saga_id: SagaId,
+    pub message: String,
+}
+
+/// Describes what happened during the last saga recovery pass
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+pub enum LastPass {
+    /// There has not been a saga recovery pass yet
+    NeverStarted,
+    /// This pass failed to even construct a plan (usually because we couldn't
+    /// load state from the database)
+    Failed { message: String },
+    /// This pass was at least partially successful
+    Success(LastPassSuccess),
```

Comment:
Should we really be calling this a "success" if it's really only sort of a success? "Planned" might be more descriptive?

Reply:
I see what you're saying, though "Planned" to me suggests "not yet executed".  "Completed"?

Reply:
:woman_shrugging: `Success` is fine with me, honestly, but `Completed` is also good. Up to you!